disease_gene/disease_associates_gene/generative_model_experiments/gen_model_benchmarking.py

```diff
@@ -0,0 +1,335 @@
+    pickle.dump(label_matricies, open("data/label_matricies.pkl", "wb"))
+
+
+# In[9]:
+
+
+lf_names = list(DG_LFS["DaG"].keys())
+
+
+# ## Visualize Label Functions
+
+# Before training the generative model, here are some visualizations for the given label functions. These visualizations are helpful in determining the efficacy of each label functions as well as observing the overlaps and conflicts between each function.
+
+# In[10]:
+
+
+plt.rcParams.update({'font.size': 10})
+plot_label_matrix_heatmap(label_matricies['train'].T, 
+                          yaxis_tick_labels=lf_names, 
+                          figsize=(10,8), font_size=10)
+
+
+# Looking at the heatmap above, this is a decent distribution of labels. Some of the label functions are covering a lot of data points (distant supervision ones) and some are very sparse in their output.
+
+# In[11]:
+
+
+plot_label_matrix_heatmap(get_overlap_matrix(label_matricies['train'], normalize=True), 
+                          yaxis_tick_labels=lf_names, xaxis_tick_labels=lf_names,
+                          figsize=(10,8), colorbar=False, plot_title="Overlap Matrix")
+
+
+# The overlap matrix above shows how two label functions overlap with each other. The brighter the color the more overlaps a label function has with another label function.
+
+# In[12]:
+
+
+plot_label_matrix_heatmap(get_conflict_matrix(label_matricies['train'], normalize=True), 
+                          yaxis_tick_labels=lf_names, xaxis_tick_labels=lf_names,
+                          figsize=(10,8), colorbar=False, plot_title="Conflict Matrix")
+
+
+# The conflict matrix above shows how often label functions conflict with each other. The brighter the color the more conflict a label function has with another function. Ignoring the diagonals, there isn't many conflicts between functions except for the LF_DG_NO_CONCLUSION and LF_DG_ALLOWED_DISTANCE.
+
+# # Train the Generative Model
+
+# After visualizing the label functions and their associated properties, now it is time to work on the generative model. As with common machine learning pipelines, the first step is to find the best hyperparameters for this model. Using the grid search algorithm, the follow parameters were optimized: amount of burnin, strength of regularization, number of epochs to run the model.
+
+# ## Set the hyperparameter grid search
+
+# In[13]:
+
+
+regularization_grid = pd.np.round(pd.np.linspace(0.1, 6, num=25), 3)
+
+
+# ## What are the best hyperparameters for the conditionally independent model?
+
+# In[37]:
+
+
+L = convert_labels(label_matricies['train'].toarray(), 'plusminus', 'categorical')
+L_dev = convert_labels(label_matricies['dev'].toarray(), 'plusminus', 'categorical')
+L_test = convert_labels(label_matricies['test'].toarray(), 'plusminus', 'categorical')
+
+validation_data = list(zip([L[:,:7], L[:, :24], L], [L_dev[:,:7], L_dev[:, :24], L_dev]))
+test_data = list(zip([L[:,:7], L[:, :24], L], [L_test[:,:7], L_test[:, :24], L_test]))
```

Comment:
These lists look a bit clunky...What do these different indices correspond to?

Reply:
These indices represent the different combination of label functions that I am training the generative model with. `:7` corresponds to the first 7 label functions, which are conveniently all knowledge based functions. The `:24` is adding the text pattern category of label functions into the model as well. I'll add some space to remove the clunky look.